MATLAB/update_FC_state.m

```diff
@@ -12,91 +12,196 @@
+        state.initialization_hold_done=true;%save this in eprom
+        state.main_state='detumble';
+    end 
+end
+if strcmp(state.main_state,'detumble')
+    %detumble
+    if norm(estimator.angular_momentum_body)<const.MAXWHEELRATE*const.JWHEEL*const.detumble_safety_factor
+        state.main_state='get gps';
+        %on real sat send command to condition magnetometers here
+    end 
+end
+return2detumble= estimator.tumbling || estimator.low_power;
+if strcmp(state.main_state,'get gps')
+    %check if gps is valid
+    if return2detumble
+        state.main_state='detumble';
+    elseif estimator.valid_gps
+        state.main_state="calibrate magnetometer 1";
+    end
+    %calculate pointing strategy
+    state.primary_current_direction_body= estimator.sat2sun_body/norm(estimator.sat2sun_body);
+    state.primary_desired_direction_body= [sqrt(2)/2;sqrt(2)/2;0];
+    state.secondary_current_direction_body= NaN(3,1);
+    state.secondary_desired_direction_body= NaN(3,1);
+end
+return2detumble= return2detumble || estimator.eclipse || ~estimator.valid_gps;
+for i=1:3
+    if strcmp(state.main_state,"calibrate magnetometer "+i)
+        axis= zeros(3,1);
+        axis(i)= 1;
+        if(estimator.sat2sun_body(i)>cos(20*pi/180))
+            state.num_magnetometer_bias_readings(i)= state.num_magnetometer_bias_readings(i)+1;
         end
-    otherwise
-        actuators.magrod_moment= [0;0;0;];
-        actuators.wheel_torque = [100;100;100;];
-        actuators.wheel_enable=[0;0;0;];
+        if return2detumble
+            state.main_state='detumble';
+        elseif state.num_magnetometer_bias_readings(i)>=const.magnetometer_bias_readings_min
+            if i<3
+                state.main_state="calibrate magnetometer "+(i+1);
+            else
+                state.main_state= 'get target orbit';
+            end
+        end 
+        %calculate pointing strategy
+        state.primary_current_direction_body= estimator.sat2sun_body/norm(estimator.sat2sun_body);
+        state.primary_desired_direction_body= axis; 
+        state.secondary_current_direction_body= NaN(3,1);
+        state.secondary_desired_direction_body= NaN(3,1);
+    end   
+end
+
+if strcmp(state.main_state,'get target orbit')
+    if return2detumble
+        state.main_state='detumble';
+    elseif estimator.valid_target_gps
+        state.main_state='rendezvous';
+    end 
+    %calculate pointing strategy
+    r_hat_eci= estimator.position_eci/norm(estimator.position_eci);
+    r_hat_body= utl_rotateframe(estimator.quat_body_eci,r_hat_eci);
+    r_hat_cross_sat2sun_body=cross(estimator.sat2sun_body,r_hat_eci);
+    state.primary_current_direction_body= r_hat_body;
+    state.primary_desired_direction_body= [1;0;0];
+    if( norm(r_hat_cross_sat2sun_body) > sin(10*pi/180))
+        state.secondary_current_direction_body= r_hat_cross_sat2sun_body/norm(r_hat_cross_sat2sun_body);
+        state.secondary_desired_direction_body= [0;0;1];
```

Comment:
Yeah, the secondary direction is NaN by default. Also, I haven't actually tested this mode, because the cdgps is perfect right now, so this mode is skipped. This can be tested more once the gps sensor model is better and there is an orbit estimator.